_posts/2018-01-26-grpc-and-deadlines.md

```diff
@@ -0,0 +1,139 @@
+
+## Setting a deadline
+
+As a client you should always set a deadline for how long you are willing to wait for a reply from the server. Here's an example using the greeting service from our [Quick Start Guides]({{ site.baseurl }}/docs/quickstart/):
+
+**C++**
+
+
+```
+  ClientContext context;
+  time_point deadline = std::chrono::system_clock::now() + 
+      std::chrono::milliseconds(100);
+  context.set_deadline(deadline);
+```
+
+
+**GO LANG**
+
+
+```
+    clientDeadline := time.Now().Add(time.Duration(*deadlineMs) * time.Millisecond)
+    ctx, cancel := context.WithDeadline(ctx, clientDeadline)
+```
+
+
+**JAVA**
+
+
+```
+response = blockingStub.withDeadlineAfter(deadlineMs, TimeUnit.MILLISECONDS).sayHello(request);
+```
+
+
+This sets the deadline to 100ms from when the client RPC is set to when the response is picked up by the client. 
+
+
+## Checking deadlines
+
+On the server side, the server can query to see if a particular RPC is no longer wanted. Before a server starts work on a response it is very important to check if there is still a client waiting for it. This is especially important to do before starting expensive processing.
+
+**C++**
+
+
+```
+if (context->IsCancelled()) {
+  return Status(StatusCode::CANCELLED, "Deadline exceeded or Client cancelled, abandoning.");
+}
+```
+
+
+**GO LANG**
+
+
+```
+if ctx.Err() == context.Canceled {
+	return status.New(codepb.CANCELLED, "Client cancelled, abandoning.")
+```
+
+
+**JAVA**
+
+
+```
+if (Context.current().isCancelled()) {
+  responseObserver.onError(Status.CANCELLED.withDescription("Cancelled by client").asRuntimeException());
+  return;
+}
+```
+
+
+Is it useful for a server to continue with the request, when you know your client has reached their deadline? It depends. If the response can be cached in the server, it can be worth processing and caching it; particularly if it's resource heavy, and costs you money for each request. This will make future requests faster as the result will already be available.
+
+
+## Adjusting deadlines
+
+What if you set a deadline but a new release or server version causes a bad regression? The deadline could be too small, resulting in all your requests timing out with `DEADLINE_EXCEEDED`, or too large and your user tail latency is now massive. You can use a flag to set and adjust the deadline.
+
+**C++**
+
+
+```
+#include <gflags/gflags.h>
+DEFINE_int32(deadline_ms, 20*1000, "Deadline in milliseconds.");
+
+  ClientContext context;
+  time_point deadline = std::chrono::system_clock::now() + 
+      std::chrono::milliseconds(FLAGS_deadline_ms);
+  context.set_deadline(deadline);
+```
+
+
+**GO LANG**
+
+
+```
+var deadlineMs = flag.Int("deadline_ms", 20*1000, "Default deadline in milliseconds.")
+
+clientDeadline := time.Duration(*deadlineMs) * time.Millisecond
+ctx, cancel := context.WithTimeout(ctx, clientDeadline)
+```
+
+
+**JAVA**
+
+
+```
+ @Option(name="--deadline_ms", usage="Deadline in milliseconds.")
```

Comment:
does this need the google-options package?

Reply:
I don't know of any OSS google options package. Without an import I couldn't tell what library it was using.